// @applitools/ufg-client/dist/resources/fetch-resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeFetchResource = void 0;
const req_1 = require("@applitools/req");
const resource_1 = require("./resource");
const buffer_1 = require("buffer");
const create_cookie_header_1 = require("../utils/create-cookie-header");
const create_user_agent_header_1 = require("../utils/create-user-agent-header");
const abort_controller_1 = require("abort-controller");
const throat_1 = __importDefault(require("throat"));
const utils = __importStar(require("@applitools/utils"));
function makeFetchResource({ concurrency, accessToken, eyesServerUrl, apiKey, tunnelIds, retryLimit = 5, timeout = 30 * 1000, streamingTimeout = 30 * 1000, cache = new Map(), fetch, logger: mainLogger, }) {
    const req = (0, req_1.makeReq)({
        retry: {
            limit: retryLimit,
            validate: ({ error }) => !!error,
        },
        fetch,
    });
    return concurrency ? (0, throat_1.default)(concurrency, fetchResource) : fetchResource;
    async function fetchResource({ resource, settings = {}, logger = mainLogger, }) {
        var _a, _b;
        logger = logger.extend(mainLogger, { tags: [`fetch-resource-${utils.general.shortid()}`] });
        let runningRequest = cache.get(resource.id);
        if (runningRequest)
            return runningRequest;
        runningRequest = req(resource.url, {
            headers: {
                Cookie: settings.cookies && (0, create_cookie_header_1.createCookieHeader)({ url: resource.url, cookies: settings.cookies }),
                ...settings.headers,
                'User-Agent': (_a = (resource.environment && (0, create_user_agent_header_1.createUserAgentHeader)({ environment: resource.environment }))) !== null && _a !== void 0 ? _a : (_b = settings.headers) === null || _b === void 0 ? void 0 : _b['User-Agent'],
            },
            proxy: resourceUrl => {
                const { proxy, autProxy } = settings;
                if (autProxy) {
                    if (!autProxy.domains)
                        return autProxy;
                    const domainMatch = autProxy.domains.includes(resourceUrl.hostname);
                    if ((autProxy.mode === 'Allow' && domainMatch) || (autProxy.mode === 'Block' && !domainMatch))
                        return autProxy;
                }
                return proxy;
            },
            hooks: [
                handleTunneledResource({ accessToken, eyesServerUrl, apiKey, tunnelIds }),
                handleLogs({ logger }),
                handleStreaming({ timeout: streamingTimeout, logger }),
            ],
            connectionTimeout: timeout,
            fallbacks: {
                shouldFallbackCondition({ response, request }) {
                    return response.status === 403 && new URL(request.url).protocol === 'https:';
                },
                updateOptions({ options }) {
                    return { ...options, keepAliveOptions: { keepAlive: true } };
                },
            },
        })
            .then(async (response) => {
            return response.ok
                ? (0, resource_1.makeResource)({
                    ...resource,
                    value: new Uint8Array(await response.arrayBuffer()),
                    contentType: response.headers.get('Content-Type'),
                })
                : (0, resource_1.makeResource)({ ...resource, errorStatusCode: response.status });
        })
            .finally(() => cache.delete(resource.id));
        cache.set(resource.id, runningRequest);
        return runningRequest;
    }
}
exports.makeFetchResource = makeFetchResource;
function handleTunneledResource({ eyesServerUrl, apiKey, accessToken, tunnelIds, }) {
    if (!tunnelIds)
        return {};
    return {
        beforeRequest({ request }) {
            if (request.url !== 'https://exec-wus.applitools.com/handle-resource') {
                return {
                    url: 'https://exec-wus.applitools.com/handle-resource',
                    method: 'POST',
                    headers: {
                        'content-type': 'application/json',
                        'x-tunnel-ids': tunnelIds,
                        'x-eyes-server-url': eyesServerUrl,
                        'x-eyes-api-key': apiKey,
                        'x-ufg-jwt-token': accessToken,
                    },
                    body: JSON.stringify({
                        resourceUrl: request.url,
                        requestHeaders: Object.fromEntries(request.headers.entries()),
                        requestTotalTimeout: 30000,
                    }),
                };
            }
        },
        afterResponse({ response }) {
            if (response.headers.has('x-fetch-status-code')) {
                const headers = {};
                for (const [name, value] of response.headers) {
                    if (name.startsWith('x-fetch-response-header')) {
                        headers[name.replace('x-fetch-response-header-', '')] = value;
                    }
                }
                return {
                    body: response.body,
                    status: Number(response.headers.get('x-fetch-status-code')),
                    statusText: response.headers.get('x-fetch-status-text'),
                    headers,
                };
            }
        },
    };
}
function handleLogs({ logger }) {
    return {
        async beforeRequest({ request }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} will be fetched using headers`, Object.fromEntries(request.headers.entries()), `and body ${JSON.stringify(await request.clone().text())}`);
        },
        beforeRetry({ request, attempt }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} will be re-fetched (attempt ${attempt})`);
        },
        afterResponse({ request, response }) {
            logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${request.url} respond with ${response.statusText}(${response.statusText})`);
        },
        afterError({ request, error }) {
            logger === null || logger === void 0 ? void 0 : logger.error(`Resource with url ${request.url} failed with error`, error);
        },
    };
}
function handleStreaming({ timeout, logger }) {
    const controller = new abort_controller_1.AbortController();
    return {
        async beforeRequest({ request }) {
            var _a, _b;
            if ((_a = request.signal) === null || _a === void 0 ? void 0 : _a.aborted)
                return;
            (_b = request.signal) === null || _b === void 0 ? void 0 : _b.addEventListener('abort', () => controller.abort());
            return { request, signal: controller.signal };
        },
        async afterResponse({ response }) {
            const contentLength = response.headers.get('Content-Length');
            const contentType = response.headers.get('Content-Type');
            const isProbablyStreaming = response.ok && !contentLength && contentType && /^(audio|video)\//.test(contentType);
            if (!isProbablyStreaming)
                return;
            return new Promise(async (resolve) => {
                const timer = setTimeout(() => {
                    controller.abort();
                    resolve({ status: 599 });
                    logger === null || logger === void 0 ? void 0 : logger.log(`Resource with url ${response.url} was interrupted, due to it takes too long to download`);
                }, timeout);
                try {
                    resolve({ response, body: buffer_1.Buffer.from(await response.arrayBuffer()) });
                }
                catch {
                    resolve({ status: 599 });
                }
                finally {
                    clearTimeout(timer);
                }
            });
        },
    };
}
